extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTenantConfigResolver.java

```diff
@@ -377,4 +380,85 @@ private static ImmutablePathMatcher.ImmutablePathMatcherBuilder<String> addPath(
+            this.tenantConfigContexts = tenantConfigContexts;
+        }
+
+        @Override
+        public String resolve(RoutingContext context) {
+            for (var tenantContext : tenantConfigContexts) {
+                final String token = OidcUtils.extractBearerToken(context, tenantContext.oidcConfig);
+                if (token != null && !OidcUtils.isOpaqueToken(token)) {
+                    final var tokenJson = OidcUtils.decodeJwtContent(token);
+                    if (tokenJson != null) {
+
+                        final String iss = tokenJson.getString(Claims.iss.name());
+                        if (tenantContext.getOidcMetadata().getIssuer().equals(iss)) {
+                            OidcUtils.storeExtractedBearerToken(context, token);
+
+                            final String tenantId = tenantContext.oidcConfig.tenantId.get();
+                            LOG.debugf("Resolved the '%s' OIDC tenant based on the matching issuer '%s'", tenantId, iss);
+                            return tenantId;
+                        }
+                    }
+                }
+            }
+            return null;
+        }
+
+        private static TenantResolver of(Map<String, TenantConfigContext> tenantConfigContexts) {
+            var contextsWithIssuer = new ArrayList<TenantConfigContext>();
+            for (TenantConfigContext context : tenantConfigContexts.values()) {
+                if (context.oidcConfig.tenantEnabled && !OidcUtils.isWebApp(context.oidcConfig)
+                        && context.getOidcMetadata() != null
+                        && context.getOidcMetadata().getIssuer() != null) {
+                    contextsWithIssuer.add(context);
+                }
+            }
+            if (contextsWithIssuer.isEmpty()) {
+                return null;
+            } else {
+                return new IssuerBasedTenantResolver(contextsWithIssuer.toArray(new TenantConfigContext[0]));
+            }
+        }
+
+        private static void addIssuerBasedTenantResolver(ArrayList<TenantResolver> resolvers,
+                Map<String, TenantConfigContext> staticTenantsConfig, TenantConfigContext defaultTenant) {
+            Map<String, TenantConfigContext> tenantConfigContexts = new HashMap<>(staticTenantsConfig);
+            tenantConfigContexts.put(OidcUtils.DEFAULT_TENANT_ID, defaultTenant);
+            var issuerTenantResolver = IssuerBasedTenantResolver.of(tenantConfigContexts);
+            if (issuerTenantResolver != null) {
+                resolvers.add(issuerTenantResolver);
+            } else {
+                LOG.warn("The 'quarkus.oidc.select-tenants-with-issuer' configuration property is set to true, "
+                        + "but no static tenant supports this feature. To use this feature, please configure at least "
+                        + "one static tenant with discovered issuer and set either 'service' or 'hybrid' application type");
```

Comment:
```suggestion
                        + "one static tenant with the discovered or configured issuer and set either 'service' or 'hybrid' application type");
```

Reply:
@michalvavrik Thinking more about it, may be it should be a debug, since the statically configured connection can be created at the first request if it fails at startup, users would need to configure the issuer in such cases where the connections may not be stable. But warning seems too strong a level, as users may set `issuer=any`, but also we let other resolvers before to check things

Reply:
It is a warning because you choose to use issuer and it won't work unless there is at least one static tenant that could be selected this way.

If we can't rely on detection during the bean creation, it would mean we need to re-check every single time all the  static tenants if something didn't change. Unless there is some event or point we can rely. But I think that would make solution unstable.

Is there a point at which we can rely on the static config being ready for inspection?

Reply:
> since the statically configured connection can be created at the first request if it fails at startup, users would need to configure the issuer in such cases where the connections may not be stable. But warning seems too strong a level, as users may set issuer=any, but also we let other resolvers before to check things

I thought there is a retry that takes care of this. Does it mean that you can postpone config resolution for static tenatns as well? Can you give me some context whether it is safe? You need to have a clear point when the metadata are ready that we can use.

Reply:
@michalvavrik 

>It is a warning because you choose to use issuer and it won't work unless there is at least one static tenant that could be selected this way.

It only enables this option because it has costs (checking of the token is opaque, get the token claims) and also it may not be what users want to do by default. It is not a requirement that the tenants must only be resolved with the issuer.

> If we can't rely on detection during the bean creation, it would mean we need to re-check every single time all the static tenants if something didn't change.

IMHO we just log-debug that the issuer based resolution will not be activated in this case.

>Does it mean that you can postpone config resolution for static tenatns as well? 

We have an oidc-wiremock test where the connection is established at the first request, and only then the tenant context discovers the metadata.

Reply:
I run `BearerTokenOidcRecoveredTest#testOidcRecoveredWithDiscovery` and as it works this way, I'll have to tweak the issuer-based resolver. Thanks

Reply:
Should be done. Let's discuss elsewhere if there are follow-ups. Closing this thread.